### ISSUE TITLE
Feature - Bulk Messaging

### ISSUE THREAD

**Issue by brutaldev** [exceptionless/Foundatio#95]
The current implementations of messaging use message pumps with callbacks which is not as efficient as bulk requesting in some circumstances. High volume scenarios should be bulk fetching to reduce costs as well since you are charged per request.

Publish and subscribe should ideally have corresponding bulk overload/versions as well.

**Comment by niemyjski** [comment]
Yes, queues also need bulk actions. Would you be willing to submit a pr / have a discussion on this? Ideally the interfaces would just have the bulk implementations and then we have overloads that just call into the bulk versions.  I think one of the issues with bulk receiving is we have a greater chance of losing messages from the brokers if we go down with our configuration we use now (we prefer speed currently).

**Comment by ejsmith** [comment]
We definitely want batch dequeue as well. That is why we have queue message locks and timeouts.
